R/module_validate.R

```diff
@@ -0,0 +1,325 @@
+        )
+      } else {
+        TRUE
+      }
+    })
+  })
+}
+
+#' @keywords internal
+srv_module_check_shinysilenterror <- function(x, validate_shiny_silent_error = TRUE) {
+  moduleServer("check_shinysilenterror", function(input, output, session) {
+    reactive({
+      if (validate_shiny_silent_error) {
+        "NEW:: Shiny silent error was raised"
+      } else {
+        TRUE
+      }
+    })
+  })
+}
+
+#' @keywords internal
+srv_module_check_teal_data <- function(x) {
+  moduleServer("check_teal_data", function(input, output, session) {
+    reactive({
+      if (inherits(x(), "qenv.error")) { # TODO: remove qenv.error
+        details <- attr(x(), "details", exact = TRUE)
+        if (is.null(details)) {
+          c(
+            "NEW:: Error when executing the `data` module:",
+            cli::ansi_strip(x()$message),
+            "",
+            "Check your inputs or contact app developer if error persists."
+          )
+        } else {
+          tagList(
+            tags$span("NEW:: Error when executing the", tags$code("data"), "module:"),
+            tags$blockquote(tags$em(cli::ansi_strip(details$condition_message))),
+            tags$span("from code:"),
+            tags$code(class = "code-error", details$current_code)
+          )
+        }
+      } else if (!inherits(x(), c("teal_data", "error"))) {
+        tags$span(
+          "NEW:: Did not receive", tags$code("teal_data"), "object. Cannot proceed further."
+        )
+      } else {
+        TRUE
+      }
+    })
+  })
+}
+
+#' @keywords internal
+srv_module_check_condition <- function(x) {
+  moduleServer("check_error", function(input, output, session) {
+    reactive({ # shiny.silent.errors are handled in a different module
+      if (inherits(x(), "error")) {
+        tagList(
+          tags$span("NEW:: Error detected:"),
+          tags$blockquote(tags$em(trimws(x()$message)))
+        )
+      } else {
+        TRUE
+      }
+    })
+  })
+}
+
+#' @keywords internal
+srv_module_check_previous_state_warn <- function(x, show_warn = reactive(FALSE), message_warn = "not defined") {
+  assert_reactive(show_warn)
+  checkmate::assert(
+    checkmate::check_string(message_warn),
+    checkmate::check_class(message_warn, "shiny.tag"),
+    checkmate::check_class(message_warn, "shiny.tag.list")
+  )
+
+  attr(message_warn, "is_warning") <- TRUE
+  moduleServer("check_shinysilenterror", function(input, output, session) {
+    reactive(if (show_warn()) message_warn else TRUE)
+  })
+}
+
+module_validate_teal_module <- module_validate_factory(
+  srv_module_check_previous_state_warn,
+  # Validate_error
+  srv_module_check_shinysilenterror,
+  srv_module_check_validation_error,
+  srv_module_check_condition,
+  srv_module_check_reactive,
+
+  srv_module_check_teal_data,
+  srv_module_check_datanames
+)
+
+module_validate_datanames <- module_validate_factory(
+  srv_module_check_previous_state_warn,
+  srv_module_check_datanames
+)
```

Comment:
There are basically 2 types of data validation groups throughout `teal`

We could reduce complexity of code and define the 2 `module_validate_xxx` manually while still keeping with low-level `srv_module_check_***` functions.